modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1160 @@
+  {
+    elem_it = elem_begin;
+    for (unsigned int j = 0; j < corner_elem_id[i]; ++j)
+      ++elem_it;
+    Elem * cut_elem = *elem_it;
+
+    for (unsigned int j = 0; j < _cut_elem_nnode; ++j)
+    {
+      bool is_edge_inside = 0;
+
+      unsigned int node1 = cut_elem->get_node(j)->id();
+      unsigned int node2 = cut_elem->get_node((j + 1 <= 2) ? j + 1 : 0)->id();
+      if (node1 > node2)
+        std::swap(node1, node2);
+
+      MeshBase::element_iterator elem_it2 = elem_begin;
+      int counter = -1;
+      for (; elem_it2 != elem_end; ++elem_it2)
+      {
+        ++counter;
+
+        if (counter != corner_elem_id[i])
+        {
+          Elem * cut_elem2 = *elem_it2;
+          for (unsigned int k = 0; k < _cut_elem_nnode; ++k)
+          {
+            unsigned int node3 = cut_elem2->get_node(k)->id();
+            unsigned int node4 = cut_elem2->get_node((k + 1 <= 2) ? k + 1 : 0)->id();
+            if (node3 > node4)
+              std::swap(node3, node4);
+
+            if (node1 == node3 && node2 == node4)
+            {
+              is_edge_inside = 1;
+              goto endloop;
+            }
+          }
+        }
+      }
+    endloop:
+      if (is_edge_inside == 0)
+      {
+        // store good edges
+        std::shared_ptr<CutEdge> ptr(new CutEdge);
+        _boundary_edges.push_back(ptr);
+        _boundary_edges[_boundary_edges.size() - 1]->id1 = node1;
+        _boundary_edges[_boundary_edges.size() - 1]->id2 = node2;
+      }
+      else
+      {
+        // this is a bad edge; remove it from existing edge list
+        for (unsigned int k = 0; k < _boundary_edges.size(); ++k)
+        {
+          if (_boundary_edges[k]->id1 == node1 && _boundary_edges[k]->id2 == node2)
+          {
+            _boundary_edges.erase(_boundary_edges.begin() + k);
+            break;
+          }
+        }
+      }
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::sortBoundaryNodes()
+// sort boundary nodes to be in the right order along the boundary
+{
+  _boundary.clear();
+
+  for (unsigned int i = 0; i < _boundary_edges.size(); ++i)
+  {
+    unsigned int node1 = _boundary_edges[i]->id1;
+    unsigned int node2 = _boundary_edges[i]->id2;
+
+    mooseAssert(_boundary_map.find(node1) != _boundary_map.end(),
+                "_boundary_map does not have this key")
+        mooseAssert(_boundary_map.find(node2) != _boundary_map.end(),
+                    "_boundary_map does not have this key")
+
+            _boundary_map.find(node1)
+                ->second->neighbor.push_back(node2);
+    _boundary_map.find(node2)->second->neighbor.push_back(node1);
+  }
+
+  std::map<int, std::shared_ptr<CutPoint>>::iterator it = _boundary_map.begin();
+  while (it != _boundary_map.end())
+  {
+    if (it->second->neighbor.size() != 2)
+      mooseError("Boundary points in cutter mesh must have exactly two neighbors");
+    ++it;
+  }
+
+  unsigned int node1 = _boundary_edges[0]->id1;
+  unsigned int node2 = _boundary_edges[0]->id2;
```

Comment:
dof_id_type - I might be missing a few of these you are using unsigned ints all over the place.